## Task 1. Transcription and Translation Analysis

### Answer:
Met-Val-Ile-Gln-Tyr

### Brief Explanation
1. 自行撰寫的c++程式協助:
	- 轉錄
	- 找出 start condon : AUG
	- 三個一組切分須轉譯的 RNA 方便查表
2. 最後查表得到轉譯後的結果

``` cpp
#include <iostream>
using namespace std;

int main()
{

    string DNA = "ACTTGTCATGGTAACTCCGTCGTACCAGTAGGTCATG";
    string RNA = "";
    
    // transcript DNA to mRNA
    for (char c : DNA)
    {
        switch (c)
        {
        case 'A':
            cout << "U";
            RNA += "U";
            break;
        case 'T':
            cout << "A";
            RNA += "A";
            break;
        case 'C':
            cout << "G";
            RNA += "G";
            break;
        case 'G':
            cout << "C";
            RNA += "C";
            break;
        default:
            break;
        }
    }
    cout << endl;
    
    // Find 'AUG' codon
    int indx  = 0;
    for (int i = 0; i < RNA.length(); i++) 
    {
        if (RNA[i] == 'A')
        {
            if (RNA[i + 1] == 'U')
            {
                if (RNA[i + 2] == 'G')
                {
	                indx = i;
                    cout << i << endl;
                }
            }
        }
    }
    
    // RNA fragment to be translated
    for (int i = indx; i < RNA.length(); i += 3) 
    {
        cout << RNA[i] << RNA[i+1] << RNA[i+2] << ' ';
    }

    return 0;
}
```
### Output: 
```
UGAACAGUACCAUUGAGGCAGCAUGGUCAUCCAGUAC
22
AUG GUC AUC CAG UAC
```

<div style="break-after: page; page-break-after: always;"></div>

## Task2. Exploring the Start of Protein Synthesis
### Protein Sequences
1. [BRCA1 Homo sapiens Protein NCBI](https://www.ncbi.nlm.nih.gov/protein/AAC37594.1?report=fasta&log$=seqview)
		![[Pasted image 20240422154612.png]]
2. [white Drosophila melanogaster Protein NCBI](https://www.ncbi.nlm.nih.gov/protein/AAF45826.1?report=fasta)
		![[Pasted image 20240422154649.png]]
3. [lacZ Escherichia coli Protein NCBI](https://www.ncbi.nlm.nih.gov/protein/CAD6021054.1?report=fasta)
		![[Pasted image 20240422154719.png]]
4. [HO Saccharomyces cerevisiae Protein NCBI](https://www.ncbi.nlm.nih.gov/protein/CAA98806.1?report=fasta)
		![[Pasted image 20240422154807.png]]
5. [FT Arabidopsis thaliana Protein NCBI](https://www.ncbi.nlm.nih.gov/protein/BAA77839.1?report=fasta)
		![[Pasted image 20240422154827.png]]

### Identify & Analyze the First Amino Acid
- 這些蛋白質序列的開頭都是 'M'(Met) 
- 參考 Chatgpt 與 [Methionine | C5H11NO2S | CID 6137 - PubChem (nih.gov)](https://pubchem.ncbi.nlm.nih.gov/compound/Methionine)
- Met 轉譯過程由 AUG 構成，AUG 是許多生物的起始密碼子。
- 甲硫胺酸 (MET) 是含有硫的非極性胺基酸，在蛋白質中有重要作用。在生物體內主要參與蛋白質合成以及其他生物化學反應中的調節。
<div style="break-after: page; page-break-after: always;"></div>

## Task 3. Codon Usage Bias Analysis
### Data reference & Tool
- [Glutamine Codon Usage and polyQ Evolution in Primates Depend on the Q Stretch Length](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5841385/)
- [Codon usage bias and dinucleotide preference in 29 Drosophila species](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8496323/)
- [Codon usage in twelve species of Drosophila](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2213667/)
- [Control of translation efficiency in yeast by codon–anticodon interactions](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2995412/)
- [Codon usage bias(wikipedia)](https://en.wikipedia.org/wiki/Codon_usage_bias)
- [General Rules for Optimal Codon Choice - PMC (nih.gov)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2700274/)
- 蒐集以上資料，使用 Microsoft copilot 進行重點整理與總結
### Summary
Codon usage bias (CUB) is a fascinating aspect of genetic code expression, where certain codons are preferred over others for the same amino acid, potentially influencing gene expression levels, protein folding, and the efficiency of protein synthesis. In the case of glutamine (Q), which is encoded by the codons CAA and CAG, a comparative analysis across Drosophila melanogaster, Homo sapiens, and Saccharomyces cerevisiae can reveal interesting evolutionary adaptations. For instance, Drosophila may exhibit a preference for CAG in highly expressed genes to optimize translation efficiency, reflecting its rapid life cycle and protein production demands. In contrast, Homo sapiens might show a more balanced use of CAA and CAG, possibly due to a more complex regulation of gene expression and protein synthesis associated with longer lifespan and more varied environmental interactions. Saccharomyces cerevisiae, being a unicellular organism, might display a distinct pattern of codon usage that correlates with its cellular machinery and metabolic requirements. Understanding these patterns can provide insights into the translational machinery's optimization in different organisms and the evolutionary pressures shaping these preferences. Moreover, variations in codon usage can affect the accuracy of protein synthesis, with potential implications for the organism's fitness and adaptability. By analyzing codon usage bias, researchers can gain valuable information on the molecular evolution of species and the functional consequences of genetic code expression. This analysis not only enhances our understanding of the genetic and evolutionary strategies of organisms but also has practical applications in biotechnology, such as optimizing gene sequences for heterologous protein expression.

<div style="break-after: page; page-break-after: always;"></div>

## Task4. Implementation of the Smith-Waterman Algorithm

## Environment
- C++ (C++14)

> [!info] Note
>  <font color="0e939f">使用 VLA(Variable Length Arrays)(C99)
>  int n = sA.size(); 
>  int m = sB.size();
>  int dp[n+1][m+1];</font>

## Code
```cpp
#include <iostream>  
#include <unordered_map>  
#include <algorithm>  
  
using namespace std;  
  
unordered_map<string, int> score{  
    {"match", 2},  
    {"mismatch", -1},  
    {"gap", -2}  
};  
  
int main(){  
    //this is a Smith-Waterman Algorithm implementation  
    string sA = "AGTCTTGACCTCAGGAGAT";  
    string sB = "CTTGACCTCAGGAGATGG";  
  
    int n = sA.size();  
    int m = sB.size();  
    int dp[n+1][m+1];  
  
    //initialize the dp table  
    for(int i = 0; i <= n; i++){  
        dp[i][0] = 0;  
    }  
    for(int i = 0; i <= m; i++){  
        dp[0][i] = 0;  
    }  
  
    //fill the dp table  
    for(int i = 1; i <= n; i++){  
        for(int j = 1; j <= m; j++){  
            int if_match = dp[i-1][j-1];  
  
            if (sA[i-1] == sB[j-1])
                if_match += score["match"];
            else
	            if_match += score["mismatch"];
            int gapA = dp[i-1][j] + score["gap"];  
            int gapB = dp[i][j-1] + score["gap"];  
            dp[i][j] = max(0, max(if_match, max(gapA, gapB)));  
        }  
    }  
  
    int maxScore = 0;  
    int x = 0;  
    int y = 0;  
  
    //find the maximum score  
    for(int i = 1; i <= n; i++){  
        for(int j = 1; j <= m; j++){  
            if(dp[i][j] > maxScore){  
                maxScore = dp[i][j];  
                x = i;  
                y = j;  
            }  
        }  
    }  
  
    string ans = "";  
  
    //trace back the path  
    while(dp[x][y] != 0){  
        int match = dp[x-1][y-1] + score["match"];  
        int mismatch = dp[x-1][y-1] + score["mismatch"];  
        int gapA = dp[x-1][y] + score["gap"];  
        int gapB = dp[x][y-1] + score["gap"];  
        if(dp[x][y] == match){  
            ans += sA[x-1];  
            x--;  
            y--;  
        }else if(dp[x][y] == mismatch){  
            ans += sA[x-1];  
            x--;  
            y--;  
        }else if(dp[x][y] == gapA){  
            ans += sA[x-1];  
            x--;  
        }else if(dp[x][y] == gapB){  
            ans += sB[y-1];  
            y--;  
        }  
    }  
    reverse(ans.begin(), ans.end());  
    cout << "The maximum score is: " << maxScore << endl;  
    cout << "The alignment is: " << ans << endl;  
    return 0;  
}
```

### Brief Explanation
1. Tools:
	- Github Copilot
	- [Teaching - Smith-Waterman (uni-freiburg.de)](https://rna.informatik.uni-freiburg.de/Teaching/index.jsp?toolName=Smith-Waterman)

2. Process:
	- Github Copilot 自動生成程式後，檢查正確性，發現 "fill the dp table" 部分程式碼有誤，原程式如下:
			
		```cpp
		for(int i = 1; i <= n; i++){
			for(int j = 1; j <= m; j++){
				// match/mismatch 計算邏輯不正確
				int match = dp[i-1][j-1] + score["match"];
				int mismatch = dp[i-1][j-1] + score["mismatch"];
				int gapA = dp[i-1][j] + score["gap"];
				int gapB = dp[i][j-1] + score["gap"];
					dp[i][j] = max(0, max(match, max(mismatch, max(gapA, gapB))));
			}
		}
		```

    - 修正後於 [Teaching - Smith-Waterman (uni-freiburg.de)](https://rna.informatik.uni-freiburg.de/Teaching/index.jsp?toolName=Smith-Waterman)驗證答案正確性
	-  ```
		The maximum score is: 32
		The alignment is: CTTGACCTCAGGAGAT
		```
    - ![[Pasted image 20240415205009.png]]
    - ![[Pasted image 20240415205030.png]]

<div style="break-after: page; page-break-after: always;"></div>

## Task 5. Multiple Sequence Alignment
### Tool
- [Clustal Omega < EMBL-EBI](https://www.ebi.ac.uk/jdispatcher/msa/clustalo/summary?jobId=clustalo-I20240423-081441-0814-25231609-p1m)
- [Clustalw Tutorial](http://ukko.life.nctu.edu.tw/~u0417082/present.html)
### Results
- ![[Pasted image 20240421213346.png]]
- ![[Pasted image 20240421223945.png]]
- ![[Pasted image 20240428192716.png]]
- ![[Pasted image 20240428192733.png]]